In [2]:
import numpy as np

In [3]:
#Define the function you want to integrate
def f(x):
    return np.exp(-x)*np.sin(x)

#Defining Simpson's 1/3 Rule
def Simpson13(f,a,b,n):
    h = ((b - a)/(n))    #step size

    #Give an intial value to what the answer is going to be
    #It is the following because the first and the last values are not multiplied by any value
    integration = f(a) + f(b)


    #Creating a for loop to figure out what terms get multiplied by what value
    #For the odd points:
    for i in range(1,n,2):
        integration += 4.0 * f(a + i * h)

    #For the even points
    for i in range(1,n-1,2):
        integration += 2.0 * f(a + i * h)

    integration *= h / 3.0

    return integration

lowLim = 0
upLim = 10
subInt = 160

#Check if the number of sub-intervals is even
if subInt % 2.0 != 0:
    print("The number of sub-intervals must be divisible by 2 for Simpson's 1/3 Rule.")
else:
    result = Simpson13(f,lowLim,upLim,subInt)

    print("The 1 dimensional approximation using Simpson's 1/3 Rule: ", result)

The 1 dimensional approximation using Simpson's 1/3 Rule:  0.5006823258721649


Lower Limit is equal to 0
Upper Limit is equal to 10
- subInt = 10 the value is 0.6215070606445476
- subInt = 20 the value is 0.5391444063425603
- subInt = 40 the value is 0.5102953529792413
- subInt = 80 the value is 0.5026269080467027
- subInt = 160 the value is 0.5006823258721649

The true value is 0.5000313962

Therefore, the error vector would be
[0.12147566, 0.03911301, 0.01026395, 0.00259551, 0.00065092]


In [4]:
ErrorVec = [0.12147566, 0.03911301, 0.01026395, 0.00259551, 0.00065092]
def GetConvergence(vec,factor=2):
    for i in range(len(vec)-1):
        print("Convergence = ", (np.log(vec[i]/vec[i+1]))/np.log(factor))
GetConvergence(ErrorVec)

Convergence =  1.6349468007051033
Convergence =  1.9300625175055044
Convergence =  1.9834960963063233
Convergence =  1.9954658982703912


In [5]:
#Define the function you want to integrate
def f(x):
    return np.exp(-x)*np.sin(x)

#Defining Simpson's 3/8 Rule
def Simpson38(f,a,b,n):
    h = ((b - a)/(n))   #step size

    #Give an intial value to what the answer is going to be
    #It is the following since the first and the last values are not multiplied by any value
    integration = f(a) + f(b)

    #Creating a for loop to figure out what terms get multiplied by what value
    for i in range(1,n):
        #if i is divisible by 3 then do the following
        if i%3 == 0:
            integration += 2 * f(a + i*h)
        #if i is not divisible by 3 do the following
        else:
            integration += 3 * f(a + i*h)

    #Multiply the integration value by 3h/8
    integration = integration * 3 * h / 8

    return integration

lowLim = 0
upLim = 10
subInt = 144

#Check to make sure that the number of sub-intervals is divisible by three
if subInt % 3 != 0:
    print("The number of sub-intervals has to be divisible by 3")
#If the number of sub-intervals is divisible by three, call the Simpson38 function
else:
    result = Simpson38(f,lowLim, upLim, subInt)
    print("The 1 dimensional approximation using Simpson's 3/8 Rule is: ", result)


The 1 dimensional approximation using Simpson's 3/8 Rule is:  0.5000308133654675


Lower Limit is equal to 0
Upper Limit is equal to 10
- subInt = 9 the value is 0.45445742501114317
- subInt = 18 the value is 0.49735470544359833
- subInt = 36 the value is 0.4998772727317998
- subInt = 72 the value is 0.5000220081513992
- subInt = 144 the value is 0.5000308133654676

The true value is 0.5000313962

Therefore, the error vector would be
[.0455739712, .0026766908, .0001541234682, .00000938804861, .00000058283454]

In [6]:
ErrorVec2 = [.0455739712, .0026766908, .0001541234682, .00000938804861, .00000058283454]
def GetConvergence(vec,factor=2):
    for i in range(len(vec)-1):
        print("Convergence = ", (np.log(vec[i]/vec[i+1]))/np.log(factor))
GetConvergence(ErrorVec2)

Convergence =  4.089687689130098
Convergence =  4.118292033398068
Convergence =  4.037117433885236
Convergence =  4.009667029763852
